### PR TITLE
improved pattern implementation and chat view

### PR DESCRIPTION
**Pull request by tmbo**
**Proposed changes**:
Rasa changes:
- improved chat view including flow diagram and e2e test
Bildschirm­foto 2023-08-21 um 09 47 19

- added an endpoint to expose flows
- allow using jinja for template formatting in response templates to allow access to nested properties. can be used as a meta property:
```yml
  utter_flow_continue_interrupted: 
    - text: Let's continue with the topic {{ flow_context.previous_flow_name }}.
      metadata: 
        rephrase: True
        template: jinja
```
DM changes:
- handling of patterns while being in patterns (cancel inside a cancel)
- removed various slots `PREVIOUS_FLOW_SLOT`, `CANCELLED_FLOW_SLOT`, `CORRECTED_SLOTS_SLOT` and replaced them with a single `FLOW_CONTEXT_SLOT`(`flow_context`) that gets set at the start of the flow. the context of a flow is stored on the stack and used whenever the flow gets active
- added default actions for cancel and correction to move the actual logic of canceling / correcting to these actions. this allows for confirmation if needed
- removed `CORRECTION` stack frame type as its usage can be replaced with the more general `REMARK`

### CODE REVIEW

File: rasa/core/actions/flows.py [hunk 1]
```diff
@@ -73,3 +80,84 @@ async def run(
...
+
+        for frame_idx in canceled_flow_frames:
+            if frame_idx >= len(stack.frames):
+                structlogger.warning(
+                    "action.cancel_flow.frame_not_found",
+                    stack=stack,
+                    frame_idx=frame_idx,
+                )
+                continue
+            stack.frames[frame_idx].step_id = END_STEP
```
Comment (twerkmeister) [inline rasa/core/actions/flows.py:121]: I don't like that we are referencing the frames using only their index here. I think there's potential for stuff to happen in between the cancel frame was put on the stack and this cancel action being executed ~ which could jumble everything.. Could stack frames get a more resilient ID?

File: rasa/core/actions/flows.py [hunk 2]
```diff
@@ -73,3 +80,84 @@ async def run(
...
+        domain: "Domain",
+        metadata: Optional[Dict[Text, Any]] = None,
+    ) -> List[Event]:
+        """Correct the slots."""
+        stack = FlowStack.from_tracker(tracker)
+        if stack.is_empty():
+            structlogger.warning("action.correct_flow_slot.no_active_flow", stack=stack)
+            return []
+
+        context = tracker.get_slot(FLOW_CONTEXT_SLOT) or {}
```
Comment (twerkmeister) [inline rasa/core/actions/flows.py:151]: same as above ~ isn't the context part of the stack frame and not a global slot?

File: rasa/core/nlg/interpolator.py [hunk 3]
```diff
@@ -52,29 +53,67 @@ def interpolate_text(response: Text, values: Dict[Text, Text]) -> Text:
...
         values: A dictionary of keys and the values that those
             keys should be replaced with.
+        method: The method to use for interpolation. If `None` or `"format"`,
 
     Returns:
         The response with any replacements made.
     """
+    if method is None or method == "format":
+        interpolator = interpolate_format
+    elif method == "jinja":
```
Comment (twerkmeister) [inline rasa/core/nlg/interpolator.py:102]: Should these method strings be constants somewhere?

File: rasa/core/nlg/response.py [hunk 4]
```diff
@@ -101,6 +102,9 @@ def _fill_response(
         # Getting the slot values in the response variables
         response_vars = self._response_variables(filled_slots, kwargs)
 
+        # template formatting method
+        method = response.get(METADATA, {}).get("template", "format")
```
Comment (twerkmeister) [inline rasa/core/nlg/response.py:106]: potential use for constants

File: rasa/core/policies/default_flows.yml [hunk 5]
```diff
@@ -56,29 +60,31 @@ flows:
         action: utter_flow_continue_interrupted
   
   pattern_correction:
-    description: A meta flow that should be started to correct a previous user input.
+    description: Confirm a previous correction of a slot value.
```
Comment (twerkmeister) [inline rasa/core/policies/default_flows.yml:63]:
```suggestion
    description: Confirm a correction of a slot value.
```

File: rasa/core/policies/flow_policy.py [hunk 6]
```diff
@@ -513,6 +495,12 @@ def _select_next_action(
                     "to __start__ if it ended it should be popped from the stack."
                 )
 
+            if top_frame := self.flow_stack.top():
+                if tracker.get_slot(FLOW_CONTEXT_SLOT) != top_frame.context:
+                    tracker.update_with_events(
+                        [SlotSet(FLOW_CONTEXT_SLOT, top_frame.context)], self.domain
+                    )
```
Comment (twerkmeister) [inline rasa/core/policies/flow_policy.py:501]: ah okay, I see that you are updating the flow context slot regularly here ~ Can you help me understand why this is necessary as opposed to using the context directly from the top most stack frame?

File: rasa/core/actions/flows.py (hunk 1 above)
Comment (twerkmeister) [reply rasa/core/actions/flows.py:121]: Also, the warning here doesn't seem to help to resolve the issue. It's indicating that the stack frame has changed, and still we are continuing with the other indices to manipulate stack frames 😬

**Review by twerkmeister — CHANGES_REQUESTED** [review]
Fantastic changes, really good job, I think this will meaningfully improve some existing architectural weaknesses.

Note: I didn't review the `chat.html` as it's more of an internal tool and we are using it, so any oddities should surface soon enough.

File: rasa/core/policies/flow_policy.py (hunk 6 above)
Comment (tmbo) [reply rasa/core/policies/flow_policy.py:501]: mhm I need to set it into a slot to have it be accessible easily from actions & templates. theoretically, I'd be also available through the stack slot, but i don't think using `flow_stack[0].context` somewhere is either save or pretty 😂 

so rather than accessing it there, I thought I'd be better to set it to a separate slot every time the flow policy advances the flows

File: rasa/core/actions/flows.py (hunk 2 above)
Comment (tmbo) [reply rasa/core/actions/flows.py:151]: yes it is, let's keep the discussion on the one comment below :+1:

File: rasa/cdu/flow_stack.py [hunk 7]
```diff
@@ -206,6 +234,8 @@ class FlowStackFrame:
     """The ID of the current step."""
     frame_type: StackFrameType = StackFrameType.REGULAR
     """The type of the frame. Defaults to `StackFrameType.REGULAR`."""
+    context: Optional[Dict[Text, Any]] = None
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/cdu/flow_stack.py:234]: I agree but I am not sure how we'd do it: we can't do it based on the frame type as e.g. most patterns are `REMARK` but have different context properties. not sure how to do this better - we could create context dataclasses and persist / parse them into context which can help catch errors, maybe - just a thought

File: rasa/cdu/command_processor.py [hunk 8]
```diff
@@ -242,22 +265,30 @@ def filled_slots_for_active_flow(
...
+                asked_questions.add(q.question)
+
+        if frame.frame_type in STACK_FRAME_TYPES_WITH_USER_FLOWS:
+            # as soon as we hit the first stack frame that is a "normal"
+            # user defined flow we stop looking for previously asked questions
+            # because we only want to ask questions that are part of the
+            # current flow.
+            continue
+
+    return asked_questions
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/cdu/command_processor.py:291]: > We break when a flow that is on the stack does not exist ~ How did this flow get on the stack in the first place?
yeah we have that in waaaaay to many places. I am very inclined to change the interface of 
```
def flow_by_id(self, id: Optional[Text]) -> Optional[Flow]:
```
to this
```
def flow_by_id(self, id: Text) -> Flow:
```
and have it throw an exception if it can't look up the id. because we really ever should have an id there that doesn't exist and if there is something is off and we can raise a full blown error.

BUT writing that: we have "mock stack frames" for intentless / search...

File: rasa/cdu/command_processor.py [hunk 9]
```diff
@@ -174,21 +178,26 @@ def execute_commands(
...
-            events.append(SlotSet(CANCELLED_FLOW_SLOT, current_top_flow.id))
+
+            canceled_frames = []
+            original_frames = FlowStack.from_dict(original_stack_dump).frames
+            for i, frame in enumerate(reversed(original_frames)):
+                # Setting the stack frame to the end step so it is properly
+                # wrapped up by the flow policy
+                canceled_frames.append(len(original_frames) - i - 1)
+                if frame.flow_id == user_flow.id:
+                    break
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/cdu/command_processor.py]: yes in a very mysterious way 😂 but if we add ids to the frames that will be more readable as well

File: rasa/cdu/command_processor.py [hunk 10]
```diff
@@ -144,20 +127,41 @@ def execute_commands(
...
+                frame_type=StackFrameType.CORRECTION,
+                context=context,
             )
+            if (
+                not current_top_flow
+                or current_top_flow.id != FLOW_PATTERN_CORRECTION_ID
+            ):
+                flow_stack.push(correction_frame)
+            else:
+                # wrap up the previous correction flow
```
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/cdu/command_processor.py:156]: the `not current_top_flow` is just there to make the type checker happy. but to be fair, could happen in practice if for some reason we end up with a correction command from somewhere.

but you are right, I can treat the no top flow case separately and noop.

File: rasa/cdu/flow_stack.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (twerkmeister) [reply rasa/cdu/flow_stack.py:234]: also just a thought ~ Could each pattern have it's own stack frame type and some stack frame types are grouped together under intermediate classes like `Regular` and `Remark`.

File: rasa/core/actions/flows.py (hunk 1 above)
Comment (twerkmeister) [reply rasa/core/actions/flows.py:121]: 🙏 yeah that should do the trick

File: rasa/core/policies/flow_policy.py (hunk 6 above)
Comment (twerkmeister) [reply rasa/core/policies/flow_policy.py:501]: Also fair points, could we have some shorthand method for that ~ `.get_flow_context()` on the stack or the tracker or something?

Comment (tmbo) [reply rasa/core/policies/flow_policy.py:501]: we could and that is a good idea for the actions + anywhere else needed.

i guess for the template we need to workaround and explicitly add it to the context passed in. but i think that is also fair.

i am good with that suggestion, removes one slot and also reduces the amounts of events we generate (which is a lot 😂 ).

File: rasa/cdu/flow_stack.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/cdu/flow_stack.py:234]: that is interesting yeah that could work - great idea. 

i'll create a ticket for it and tackle it separately, don't want to hold this pr for too long.

[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/cdu/flow_stack.py:234]: https://rasahq.atlassian.net/browse/ENG-499

File: rasa/cdu/command_processor.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/cdu/command_processor.py:291]: I just double checked, the `continue` should be a `break` - good catch :+1:

[the remark this answers is not in the document]
Comment (tmbo) [reply rasa/cdu/command_processor.py:291]: i'll keep the interface of flow_by_id like it is for now and tackle that separately.

File: rasa/core/actions/flows.py (hunk 1 above)
Comment (tmbo) [reply rasa/core/actions/flows.py]: I've changed the implementation and added ids to every stackframe.

File: rasa/core/policies/flow_policy.py (hunk 6 above)
Comment (tmbo) [reply rasa/core/policies/flow_policy.py]: Changed it: we do not have a separate slot for the context. the actions directly pull it from the stack and in response templates i inject the context as a variable so that it can be used there.

File: rasa/cdu/command_processor.py [hunk 11]
```diff
@@ -144,20 +127,47 @@ def execute_commands(
...
+                flow_stack.frames[-1].step_id = END_STEP
+                # push a new correction flow
+                flow_stack.push(
+                    correction_frame,
+                    # we allow the previous correction to finish first before
+                    # starting the new one
+                    index=-1,
+                )
```
Comment (twerkmeister) [inline rasa/cdu/command_processor.py:170]: Actually, I think this would allow for attacking the flow diagram step logic. An "attack" could look like this:
1. go through the flow with values you think will satisfy conditions and get you to a desired leaf node
2. introduce some odd corrections (e.g. more money than you have, negative monetary value or sth)
3. introduce another correction before accepting the first one that is closest to your desired leaf node. 

Step two would introduce the bad data, and jump back. Before the steps are re-executed, step three would jump you back close to your desired leaf note, potentially skipping a lot of validations.

---
Even if it might be good to do a final validation before some API call, it seems we do want to rely on the flow logic validation quite a bit, as it is necessary for the branching logic. So we should probably address this. One potential solution would probably be merging the two frames ~ up to you if you want to do it here or in a follow up ticket i'd say

File: rasa/cdu/command_processor.py [hunk 12]
```diff
@@ -174,21 +184,29 @@ def execute_commands(
...
-                    flow_stack.frames[idx].step_id = END_STEP
-            events.append(SlotSet(CANCELLED_FLOW_SLOT, current_top_flow.id))
+
+            canceled_frames = []
+            original_frames = FlowStack.from_dict(original_stack_dump).frames
+            # we need to go through the original stack dump in reverse order
+            # to find the frames that were canceled. we cancel everthing from
+            # the top of the stack until we hit the user flow that was canceled.
+            # this will also cancel any patterns put ontop of that user flow,
+            # e.g. corrections.
```
Comment (twerkmeister) [inline rasa/cdu/command_processor.py:194]: We should probably test what the llm does when someone tries to cancel a correction by saying "no, stop" or "cancel". Could be an e2e test case.

**Review by twerkmeister — APPROVED** [review]
Noice 🚀. Just noticed two more edge cases, but can probably be addressed later, with separate tickets

File: rasa/cdu/command_processor.py (hunk 12 above)
Comment (twerkmeister) [reply rasa/cdu/command_processor.py:194]: oh yeah, I think you are handling the "don't do the correction case". I am just wondering about the interplay between "no" and "cancel" here. If someone says cancel to the question "Do you want to update the recipient?" then they might end up cancelling the whole transfer money flow, as cancellations dig down to the user frame

File: rasa/cdu/command_processor.py (hunk 11 above)
Comment (twerkmeister) [reply rasa/cdu/command_processor.py:170]: maybe I also got it wrong as to what is currently happening here, yeah let's discuss
